src/host.cpp

```diff
@@ -185,6 +190,98 @@ namespace rhost {
         }
 
+        void terminate_if_disconnected() {
+            // terminate invokes R_Suicide, which may invoke WriteConsole and/or ShowMessage, which will
+            // then call terminate again, so we need to prevent infinite recursion here.
+            static bool is_terminating;
+            if (is_terminating) {
+                return;
+            }
+
+            if (!transport::is_connected()) {
+                is_terminating = true;
+                terminate("Lost connection to client.");
+            }
+        }
+
+        void shutdown_if_requested() {
+            terminate_if_disconnected();
+
+            if (!shutdown_requested) {
+                return;
+            }
+
+            // Make sure we don't try handle the pending shutdown request more than once.
+            static std::atomic<bool> is_shutting_down;
+            bool expected = false;
+            if (!is_shutting_down.compare_exchange_strong(expected, true)) {
+                return;
+            }
+
+            if (!rdata.empty()) {
+                std::string s = rdata.string();
+                logf(log_verbosity::minimal, "Saving workspace to %s...\n", s.c_str());
+
+                bool saved = r_top_level_exec([&] {
+                    R_SaveGlobalEnvToFile(s.c_str());
+                });
+
+                logf(log_verbosity::minimal, saved ? "Workspace saved successfully.\n" : "Failed to save workspace.\n");
+                send_notification("!End", saved);
+            }
+
+            terminate("Shutting down by request.");
+        }
+
+        void request_shutdown(bool save_rdata) {
+            if (!save_rdata) {
+                rdata.clear();
+            }
+
+            shutdown_requested = true;
+
+            std::thread([] {
+                std::this_thread::sleep_for(1min);
+                terminate("Timed out while waiting for graceful shutdown to complete; terminating process.");
+            }).detach();
+
+            unblock_message_loop();
+        }
+
+        void request_shutdown(const message& msg) {
+            assert(!strcmp(msg.name(), "!Shutdown"));
+
+            if (shutdown_requested) {
+                return;
+            }
+
+            auto json = msg.json();
+            if (!json[0].is<bool>()) {
+                fatal_error("Invalid evaluation request: 1 boolean argument expected");
+            }
+            auto save_rdata = json[0].get<bool>();
+
+            request_shutdown(save_rdata);
+        }
+
+        void idle_timer_thread(std::chrono::seconds idle_timeout) {
+            for (;;) {
+                std::chrono::steady_clock::time_point idling_since;
+                {
+                    std::lock_guard<std::mutex> lock(idle_timer_lock);
+                    idling_since = host::idling_since;
+                }
+
+                auto delta = std::chrono::steady_clock::now() - idling_since;
+                if (delta > idle_timeout) {
+                    request_shutdown(true);
```

Comment:
The default is no timeout at all. This is represented by a 0-second timeout, but the background watcher thread isn't even started in this case:

https://github.com/Microsoft/R-Host/pull/91/files#diff-59b9dbd011d28180f916f3a21d1af75fR1020